src/key.rs

```diff
@@ -0,0 +1,309 @@
+        let s = CString::new(v).unwrap().into_string().unwrap();
+        writeln!(f, "{:?} {:?}:", self.get_type().unwrap(), self.get_length().unwrap())?;
+        write!(f, "{}", s)
+    }
+}
+
+impl Drop for Key {
+    fn drop(&mut self)
+    {
+        unsafe {
+            lib::yaca_key_destroy(self.handle as *mut c_void)
+        }
+    }
+}
+
+impl Key {
+    /// Generates a secure key or key generation parameters
+    /// (or an Initialization Vector).
+    pub fn generate(key_type: &KeyType, key_length: &KeyLength) -> Result<Key>
+    {
+        key_generate(key_type, key_length)
+    }
+    /// Generates a secure private asymmetric key from parameters.
+    pub fn generate_from_parameters(params: &Key) -> Result<Key>
+    {
+        key_generate_from_parameters(params)
+    }
+    /// Imports a key or key generation parameters.
+    pub fn import(data: &[u8], key_type: &KeyType, password: Option<&CStr>) -> Result<Key>
+    {
+        key_import(data, key_type, password)
+    }
+    /// Exports a key or key generation parameters to arbitrary format.
+    pub fn export(&self, key_fmt: &KeyFormat, key_file_fmt: &KeyFileFormat,
+                  password: Option<&CStr>) -> Result<Vec<u8>>
+    {
+        key_export(self, key_fmt, key_file_fmt, password)
+    }
+    /// Extracts public key from a private one.
+    pub fn extract_public(&self) -> Result<Key>
+    {
+        key_extract_public(self)
+    }
+    /// Extracts parameters from a private or a public key.
+    pub fn extract_parameters(&self) -> Result<Key>
+    {
+        key_extract_parameters(self)
+    }
+    /// Derives a shared secret using Diffie-Helmann or EC Diffie-Helmann
+    /// key exchange protocol.
+    pub fn derive_dh(prv_key: &Key, pub_key: &Key) -> Result<Vec<u8>>
+    {
+        key_derive_dh(prv_key, pub_key)
+    }
+    /// Derives a key material from shared secret.
+    pub fn derive_kdf(kdf: &Kdf, algo: &DigestAlgorithm,
+                      secret: &[u8], info: Option<&[u8]>,
+                      key_material_len: usize) -> Result<Vec<u8>>
+    {
+        key_derive_kdf(kdf, algo, secret, info, key_material_len)
+    }
+    /// Derives a key from user password (PKCS #5 a.k.a. pbkdf2 algorithm).
+    pub fn derive_pbkdf2(password: &CStr, salt: Option<&[u8]>, iterations: usize,
+                         algo: &DigestAlgorithm, key_bit_len: usize) -> Result<Key>
+    {
+        key_derive_pbkdf2(password, salt, iterations, algo, key_bit_len)
+    }
+    /// Gets key's type.
+    pub fn get_type(&self) -> Result<KeyType>
+    {
+        key_get_type(&self)
+    }
+    /// Gets key's length.
+    pub fn get_length(&self) -> Result<KeyLength>
+    {
+        key_get_length(&self)
+    }
+}
+
+// Used by the C wrappers
+#[inline]
+pub(crate) fn get_handle(key: &Key) -> *const c_void
+{
+    key.handle
+}
+
+// Used to compose Keys in Seal
+#[inline]
+pub(crate) fn new_key(handle: *const c_void) -> Key
+{
+    Key{handle}
```

Comment:
In general I'd agree, but this function is used in 2 places in one function and I would duplicate one assert anyway (from match) so no need for now. If this will be used more (it wont) then I'll change.

Reply:
Sure, np. Where I'm coming from:
* Assert everything that moves.
* Don't mind spurious/redundant assertions.

Reply:
You might want to change your points to:
* debug_assert everything that moves.
* Don't mind spurious/redundant debug_assertions.

... since we'll be working in Rust for the foreseeable future :-)

Reply:
Yeah, thanks for staying vigilant :)

"While the spirit is vigilant the body thrives"